repository.go

```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
 	return nil
 }
 
+func (r repository) applyMutates(cw contextWrapper, doc *Document, mutation Mutation, filter FilterQuery) error {
+	var (
+		baseQueries = []Querier{filter, mutation.Unscoped, mutation.Cascade}
+		queries     = baseQueries
+	)
+	if version, ok := r.lockVersion(*doc, mutation.Unscoped); ok {
+		Inc("lock_version").SkipReload().Apply(doc, &mutation)
+		queries = append(queries, LockVersion(version))
+	}
+
+	var (
+		pField string
+		query  = r.withDefaultScope(doc.data, Build(doc.Table(), queries...), false)
+	)
+
+	if len(doc.data.primaryField) == 1 {
+		pField = doc.PrimaryField()
+	}
+
+	if updatedCount, err := cw.adapter.Update(cw.ctx, query, pField, mutation.Mutates); err != nil {
+		return mutation.ErrorFunc.transform(err)
+	} else if updatedCount == 0 {
+		return NotFoundError{}
+	}
+
+	if mutation.Reload {
+		baseQuery := r.withDefaultScope(doc.data, Build(doc.Table(), baseQueries...), false)
+		if err := r.find(cw, doc, baseQuery.UsePrimary()); err != nil {
+			return err
+		}
+	} else if version, ok := r.lockVersion(*doc, mutation.Unscoped); ok {
+		doc.SetValue("lock_version", version+1)
```

Comment:
I think this will return incorrect version if multiple process is updating the same record concurrently without using lock

to make things simple, I think what we can do is:
- if lock is **enabled**: automatically update using Set with new version value is computed from rel side
- if lock is **disabled**: don't automatically update version, warn user that lock version will be replaced

Reply:
Sorry for the long pause, had to take a break.

Why? If `SET version = version + 1 WHERE version = x` exectues successfully, then the version has to be x + 1. As far as I understand, we don't care about other fields. Mixing usage with and without locks is up to the user.

Reply:
> If SET version = version + 1 WHERE version = x exectues successfully, then the version has to be x + 1.

ah right this is true

but I can see the implementation can be simplified if we just do `SET version = newVersion WHERE version = previousVersion`, where we locally compute `newVersion = previousVersion+1` in memory

this way, we don't have to introduce NoReload, semantics that rel.Inc/rel.Dec/rel.Fragment always trigger reload will remain true.
(it always trigger reload, because there's no way to compute the updated record correctly in memory)

Reply:
Yeah, thats true.

They key problem here is that Set() calls are always applied to the document beforehand (in Mutate.Apply()), regardless of whether the record is found/any erros occured. This is the default behaviour for rel now. 
Should version locks be rolled back in case of unsuccessful updates?

Reply:
ah true,

> Should version locks be rolled back in case of unsuccessful updates?

I think yes